In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Q1. Read the data for January. How many columns are there?

In [2]:
# read dataframe and take a look
df_jan = pd.read_parquet('yellow_tripdata_2022-01.parquet')


In [3]:
df_jan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2463931 entries, 0 to 2463930
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[ns]
 2   tpep_dropoff_datetime  datetime64[ns]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

In [4]:
# print the rows and columns
print(df_jan.shape) 
df_jan.head()

(2463931, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65           0.0   
1             1          8.0    0.5      0.5        4.00           0.0   
2             1          7.5    0.5      0.5        1.76           0.0   
3             2          8.0    0.5      0.5        0.00           0.0   
4             1         23.5    0.5      0.5        3.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         21.95                   2.5          0.0  
1                    0.3         13.30                   0.0          0.0  
2                    0.3         10.56                   0.0          0.0  
3                    0.3         11.80                   2.5          0.0  
4                    0.3         30.30                   2.5          0.0

## Q1. Answer

19 columns

# Q2. Computing duration

Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

In [5]:
def create_target(df: pd.DataFrame) -> pd.DataFrame:
    df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    
    df['duration'] = df['duration'].apply(lambda duration: duration.total_seconds() / 60)
    
    return df

In [6]:
# Calculating the duration
create_target(df_jan)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee   duration  
0                         2.5          0.0  17.816667  
1                         0.0          0.0   8.400000  
2                         0.0          0.0   8.966667  
3                         2.5          0.0  10.033333  
4                         2.5          0.0  37.533333  
...                       ...          ...        ...  
2463926                   NaN          NaN   5.966667  
2463927                   NaN

In [7]:
# Std
round(df_jan.duration.std(),2)

46.45

## Q2. Answer

Standard deviation is 46.45

# Q3. Dropping outliers

Next, we need to check the distribution of the duration variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

In [8]:
df_jan.duration.describe()

count    2.463931e+06
mean     1.421220e+01
std      4.644531e+01
min     -3.442400e+03
25%      6.316667e+00
50%      1.018333e+01
75%      1.616667e+01
max      8.513183e+03
Name: duration, dtype: float64

In [9]:
def filter_df(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    df = df[(df.duration >= 1) & (df.duration <= 60)]
    
    return df.reset_index(drop=True)

In [10]:
# filtering the outliers between 1 and 60
df_jan_filtred = filter_df(df_jan)

In [11]:
# Get fraction of records left
df_jan_filtred.shape[0] / df_jan.shape[0] * 100

98.27547930522405

## Q3. Anwser

The fraction is 98.27 (round - 98)

# Q4. One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

* Turn the dataframe into a list of dictionaries
* Fit a dictionary vectorizer
* Get a feature matrix from it
* What's the dimensionality of this matrix (number of columns)?

In [12]:
cat_columns = ['PULocationID', 'DOLocationID']
df_jan_filtred[cat_columns] = df_jan_filtred[cat_columns].astype(str)
train_dicts = df_jan_filtred[cat_columns].to_dict(orient='records')



dv = DictVectorizer()
X_train = dv.fit_transform(train_dicts)

In [13]:
X_train.shape

(2421440, 515)

## Q4. Anwser

515 columns

# Q5. Training a model
Now let's use the feature matrix from the previous step to train a model.

* Train a plain linear regression model with default parameters
* Calculate the RMSE of the model on the training data
* What's the RMSE on train?

In [14]:
y_train = df_jan_filtred.duration.values

lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [15]:
y_pred = lr.predict(X_train)

In [16]:
round(mean_squared_error(y_train, y_pred, squared=False),2)

6.99

## Q5. Anwser

RMSE 6.99

# Q6. Evaluating the model

Now let's apply this model to the validation dataset (February 2022).

What's the RMSE on validation?

In [17]:
df_feb = pd.read_parquet('yellow_tripdata_2022-02.parquet')
print(df_feb.shape)
df_feb.head()

(2979431, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-02-01 00:06:58   2022-02-01 00:19:24              1.0   
1         1  2022-02-01 00:38:22   2022-02-01 00:55:55              1.0   
2         1  2022-02-01 00:03:20   2022-02-01 00:26:59              1.0   
3         2  2022-02-01 00:08:00   2022-02-01 00:28:05              1.0   
4         2  2022-02-01 00:06:48   2022-02-01 00:33:07              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           5.40         1.0                  N           138           252   
1           6.40         1.0                  N           138            41   
2          12.50         1.0                  N           138           200   
3           9.88         1.0                  N           239           200   
4          12.16         1.0                  N           138           125   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         17.0   1.75      0.5        3.90          0.00   
1             2         21.0   1.75      0.5        0.00          6.55   
2             2         35.5   1.75      0.5        0.00          6.55   
3             2         28.0   0.50      0.5        0.00          3.00   
4             1         35.5   0.50      0.5        8.11          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         23.45                   0.0         1.25  
1                    0.3         30.10                   0.0         1.25  
2                    0.3         44.60                   0.0         1.25  
3                    0.3         34.80                   2.5         0.00  
4                    0.3         48.66                   2.5         1.25

In [18]:
create_target(df_feb)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-02-01 00:06:58   2022-02-01 00:19:24              1.0   
1               1  2022-02-01 00:38:22   2022-02-01 00:55:55              1.0   
2               1  2022-02-01 00:03:20   2022-02-01 00:26:59              1.0   
3               2  2022-02-01 00:08:00   2022-02-01 00:28:05              1.0   
4               2  2022-02-01 00:06:48   2022-02-01 00:33:07              1.0   
...           ...                  ...                   ...              ...   
2979426         2  2022-02-28 23:50:00   2022-03-01 00:06:00              NaN   
2979427         2  2022-02-28 23:06:57   2022-02-28 23:19:12              NaN   
2979428         2  2022-02-28 23:48:13   2022-03-01 00:03:33              NaN   
2979429         2  2022-02-28 23:56:41   2022-03-01 00:04:57              NaN   
2979430         2  2022-02-28 23:04:00   2022-02-28 23:10:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 5.40         1.0                  N           138   
1                 6.40         1.0                  N           138   
2                12.50         1.0                  N           138   
3                 9.88         1.0                  N           239   
4                12.16         1.0                  N           138   
...                ...         ...                ...           ...   
2979426           3.40         NaN               None           163   
2979427           3.48         NaN               None           141   
2979428           3.05         NaN               None           161   
2979429           2.62         NaN               None           141   
2979430           1.41         NaN               None           237   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 252             1        17.00   1.75      0.5        3.90   
1                  41             2        21.00   1.75      0.5        0.00   
2                 200             2        35.50   1.75      0.5        0.00   
3                 200             2        28.00   0.50      0.5        0.00   
4                 125             1        35.50   0.50      0.5        8.11   
...               ...           ...          ...    ...      ...         ...   
2979426           193             0        14.37   0.00      0.5        0.00   
2979427             4             0        14.51   0.00      0.5        2.00   
2979428           151             0        14.38   0.00      0.5        3.81   
2979429           226             0        12.53   0.00      0.5        1.71   
2979430           142             0         8.80   0.00      0.5        2.57   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3         23.45   
1                6.55                    0.3         30.10   
2                6.55                    0.3         44.60   
3                3.00                    0.3         34.80   
4                0.00                    0.3         48.66   
...               ...                    ...           ...   
2979426          0.00                    0.3         17.67   
2979427          0.00                    0.3         19.81   
2979428          0.00                    0.3         21.49   
2979429          0.00                    0.3         17.54   
2979430          0.00                    0.3         14.67   

         congestion_surcharge  airport_fee   duration  
0                         0.0         1.25  12.433333  
1                         0.0         1.25  17.550000  
2                         0.0         1.25  23.650000  
3                         2.5         0.00  20.083333  
4                         2.5         1.25  26.316667  
...                       ...          ...        ...  
2979426                   NaN          NaN  16.000000  
2979427                   NaN

In [19]:
df_feb_fitered = filter_df(df_feb)

In [20]:
df_feb_fitered[cat_columns] = df_feb_fitered[cat_columns].astype(str)

In [21]:
val_dicts = df_feb_fitered[cat_columns].to_dict(orient='records')

In [22]:
X_val = dv.transform(val_dicts)
print(X_val.shape)

(2918187, 515)


In [23]:
y_val = df_feb_fitered['duration'].values

In [24]:
y_pred_val = lr.predict(X_val)

In [25]:
round(mean_squared_error(y_val, y_pred_val, squared=False),2)

7.79

## Q6. Anwser
RMSE 7.79